source/agora/utils/Log.d

```diff
@@ -49,21 +49,24 @@ public struct Logger
         import core.memory;
         this.logger = Ocean.Log.lookup(moduleName);
         this.logger.buffer(new char[](16384));
-        GC.addRoot(cast(void*)this.logger);
-    }
-
-    /// dtor
-    public ~this ()
-    {
-        import core.memory;
-        GC.removeRoot(cast(void*)this.logger);
     }
 
     public void opDispatch (string call, Args...) (Args args)
     {
         try
         {
-            mixin("this.logger." ~ call ~ "(args);");
+            import core.memory : GC;
+            if (GC.inFinalizer())
+            {
+                writeln("allowing logging from the destructor on a GC thread would risk running into segfaults, please see issue #1128");
+                return;
+            }
+            else if (logger is null)
+            {
+                assert(0,"\nplease make sure you are declaring the \nmixin AddLogger!(); statement on top, followed by:\nstatic this{}; followed by:\nstatic ~this{};");
```

Comment:
Do we really want to crash here ? Might as well segv and get a core dump if we do. What's the rationale for using `assert(0)` ?

Reply:
https://github.com/bpfkorea/agora/issues/1128#issuecomment-676932589

this code corresponds to the point 1(where the static constructor/destructor is arranged in an unexpected order and causes SEGFAULTS at every startup consistently during executing the static section) -> this is the less serious issue compared to the other one. As you said I can just let it segfault, but I thought I would give a human readable error message.

The more serious issue which this PR resolves correspond to the case when we get SEGFAULTS by logging from the destructors. I plan to give more detail about that tomorrow.